src/WinRT.Runtime/TypeNameSupport.cs

```diff
@@ -187,26 +187,39 @@ private static Type FindTypeByNameCore(string runtimeClassName, Type[] genericTy
             {
                 if (genericTypes != null)
                 {
+                    return ResolveGenericType(resolvedType, genericTypes, runtimeClassName);
+                }
+                return resolvedType;
+            }
+
+            Debug.WriteLine($"FindTypeByNameCore: Unable to find a type named '{runtimeClassName}'");
+            return null;
+
 #if NET
-                    if (!RuntimeFeature.IsDynamicCodeCompiled)
+            [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Calls to MakeGenericType are done with reference types")]
+#endif
+            static Type ResolveGenericType(Type resolvedType, Type[] genericTypes, string runtimeClassName)
+            {
+                if (resolvedType == typeof(global::System.Nullable<>) && genericTypes[0].IsDelegate())
+                {
+                    return typeof(ABI.System.Nullable_Delegate<>).MakeGenericType(genericTypes);
+                }
+
+#if NET
+                foreach (var type in genericTypes)
+                {
+                    if (type.IsValueType)
                     {
-                        throw new NotSupportedException($"Cannot provide generic type from '{runtimeClassName}'.");
+                        if (!RuntimeFeature.IsDynamicCodeCompiled)

```

Comment:
Nit: can hoist the check here? Ie. just check for `IsDynamicCodeCompiled` once and skip the whole loop if we're not on NAOT. I mean technically the code is the same since it's a constant, but seems easier to read and could help the JIT on CoreCLR just a little bit more to remove the entire loop?

Reply:
good point, addressed.